### DISCUSSION TITLE
Retrieve Data From Other Services/Modules

### DISCUSSION THREAD

**Discussion by dtquocbao** [oqtane/oqtane.framework#853] [Q&A] [answered]
Hi,

So basically I am working modules that need to retrieve data which is already manipulated from other modules (kind of admin settings) and I found that Oqtane use CreateAuthorizationPolicyUrl helper function which needs to pass by the ModuleID in other to get data from services/modules. So that means we need the ModuleID all the time as expected of Oqtane. 

My question is can you guys advise any idea to handle this case?

**Comment by hishamco** [comment]
>  (kind of admin settings)

I think you could easily inject `ISiteService` in this case

**Reply by dtquocbao** [reply]
Thanks,

In this case it's my own settings.  For example I created a module named "Message Settings" which stored all my message template config. Now from module "Message Sending" I would like to load the settings of message template. However, in my opinion, there are two options:

1. I will create a new API in Message Sending to retrieve Message Settings repo. 
2. I will modify Message Settings API to retrieve data (without ModuleID?).

**Reply by hishamco** [reply]
If you register your service from the first module into DI, you will be able to retrieve the `IMessageSettings` from your second module, Oqtane will handle this through `IServerStartup` & `IClientStartup`

**Reply by dtquocbao** [reply]
Yeah, I am able to retrieve service from another service. But seems we need to pass the ModuleID to call that service with authorization.

**Reply by hishamco** [reply]
Why? you are the owner of the service, so if you register `IMessageSettings`, you will be able to resolve it from DI in another module, I'm not sure why you need AUTH here?!!

**Comment by sbwalker** [comment]
It is not clear what you are trying to accomplish. If you look at the Blog module ( https://github.com/oqtane/oqtane.blogs - which was created using the Module Creator ) you will see that it contains a "settings" component which allows you to set configuration which can be used to control the main display component. If you use this pattern then the settings component will be automatically displayed within the module settings UI. 

I will echo @hishamco's comment that you really should not have to concern yourself with Auth at all if you are using the framework as expected - Auth should be an infrastructure item which is invisible to the developer. The whole concept of Oqtane is that it is a module framework - so you build Module components rather than traditional Page components. Module components must inherit from ModuleBase. ModuleBase provides access to ModuleState - which provides a rich set of metadata including ModuleId within each component.

**Comment by hishamco** [comment]
> I will echo @hishamco's comment that you really should not have to concern yourself with Auth at all if you are using the framework 

That's what I expect, but @rockydant made me confused a little bit when he told me an AUTH is needed, while I'm sure not AUTH needed here

**Comment by dtquocbao** [comment, marked as answer]
Thanks @hishamco and @sbwalker I got what you guys are advising about and very clear with the Blog example. But what if I would like to retrieve data from other repositories instead of the current repository of each module? 

E.g: Module A(Repository A - Table A) retrieves data from Table B of Module B(Service B - Repository B). Should I 

1. Client: call API service from Service B
2. Server: reference repository B from repository A and respond to component, or modify repository A to query more data from table B

**Reply by sbwalker** [reply]
OK, I think I understand your question now, and how it related to Auth. Basically you have some back-end services which need to be shared  across a number of front-end modules. This is similar to the Oqtane framework itself, as it has a variety of back-end services which are shared across modules. In this case, your back-end services should not use the custom authorization policies ( ie. [Authorize(Policy = "ViewModule")] ) as these are dependent upon a specific ModuleId. Instead, your shared services would most likely use role-based authorization ( ie. [Authorize(Roles = RoleNames.Registered)] ). And your front-end service wrapper methods would not need the CreateAuthorizationPolicyUrl helper method, as they would not need to pass the ModuleId to the back-end.

**Reply by dtquocbao** [reply]
That's fantastic! Makes sense now @sbwalker. 

Thank you!

**Reply by sbwalker** [reply]
After thinking about this a bit more, you actually could use the custom authorization policies - however you need to ensure you have set the module Permissions appropriately for each front-end module instance on the page. As an example lets say if on a page in your site you have an instance of ModuleA (ModuleID=1) and an instance of ModuleB (ModuleID=2), and in BOTH cases you have configured the module's Permissions ( via the Module Settings UI ) so that Registered Users have View permissions. Your front-end service wrapper methods within ModuleA and ModuleB could use the CreateAuthorizationPolicyUrl helper method to pass the associated ModuleId to the back-end. And lets assume you have ServiceA as a shared back-end service API endpoint. You could specify [Authorize(Policy = "ViewModule")] on its methods. Then when either the ModuleA or ModuleB instance called ServiceA the request would contain a ModuleID and the system would lookup the View permissions for the ModuleID and determine that RegisteredUsers should have access and then verify that the user making the call is in the RegisteredUser role.

**Reply by dtquocbao** [reply]
That's cool @sbwalker 

What about for example 

- ModuleA(CRUD list of invoice statuses like new/processing/complete...) on PageA, 
- ModuleB on PageB which CRUD invoices and need to adjust the statuses of each invoice as well,
- ModuleC on PageC which uses invoice statuses to make a search module for other purposes.

In this case I guess we can't use ModuleID to load ModuleA statuses?

**Reply by sbwalker** [reply]
@rockydant absolutely you can. As long as you configure the module instances with the appropriate permissions it does not matter what page those modules are located on. Remember that that is a client/server application - the front-end and back-end are completely independent from one another. When you add a [Authorize(Policy = "ViewModule")] attribute to a back-end method you are saying that when that method is called a ModuleID will be passed on the Url which will allow the system to retrieve the permissions which have been defined for that ModuleID from the database. Those permissions will indicate which users/roles have been granted access to call the method. And just like everything else in Oqtane, the permissions are not hard-coded - they are dynamic. The easiest way to configure them is by using the Module Settings UI associated to each module instance within the front-end portal framework. And using this UI you can modify the permissions at run-time at any time without requiring a redeployment of your application or modules. You could also script the permissions in the database manually yourself but that would be a lot of extra work.

**Reply by dtquocbao** [reply]
Yeah I just wonder if it's acceptable to pass by the CreateAuthorizationPolicyUrl but gotcha now. Thanks
